File: Level1/Lotto.cpp
/*
    https://programmers.co.kr/learn/courses/30/lessons/77484
*/

#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> lottos, vector<int> win_nums) {
    const vector<int> award{ 6, 6, 5, 4, 3, 2, 1 };
    int best_score = 0;
    int worst_score = 0;

    for (int i = 0; i < lottos.size(); ++i)
    {
        if (lottos[i] == 0)
        {
            ++best_score;
            continue;
        }

        for (int j = 0; j < win_nums.size(); ++j)
        {
            if (win_nums[j] == lottos[i])
            {
                ++best_score;
                ++worst_score;
                break;
            } 
        }
    }

    return vector<int>{award[best_score], award[worst_score]};
}

File: Java/Level1/Lotto.java
/*
 * https://school.programmers.co.kr/learn/courses/30/lessons/77484
 */
import java.util.Arrays;

class Solution {
    public int[] solution(int[] lottos, int[] win_nums) {
        int[] answer = {0, 0};

        int lottoIndex = 0;
        int winNumsIndex = 0;
        int joker = 0;
        int matchCount = 0;

        Arrays.sort(lottos);
        Arrays.sort(win_nums);

        while(lottoIndex < lottos.length && winNumsIndex < win_nums.length) {
            if(lottos[lottoIndex] == 0) {
                ++joker;
                ++lottoIndex;
                continue;
            } 

            if(lottos[lottoIndex] == win_nums[winNumsIndex]) {
                ++matchCount;
                ++lottoIndex;
                ++winNumsIndex;
            } else if(lottos[lottoIndex] > win_nums[winNumsIndex]) {
                ++winNumsIndex;
            } else {
                ++lottoIndex;
            }
        }
        
        answer[0] = Math.min(6 - (matchCount + joker - 1), 6);
        answer[1] = Math.min(6 - (matchCount - 1), 6);

        return answer;
    }
}